File: javascript/simple-cipher/simple-cipher.js
const alphabet = 'abcdefghijklmnopqrstuvwxyz';
const len = alphabet.length;
const index = c => alphabet.indexOf(c);

const encoded = (c, k) => alphabet[(index(c) + index(k)) % len];
const decoded = (c, k) => alphabet[(index(c) - index(k) + len) % len];

const generateRandomKey = () => {
  const getRandomInt = max => Math.floor(Math.random() * Math.floor(max));
  let key = '';
  for (let i = 1; i <= 100; i += 1) {
    key += alphabet[getRandomInt(len)];
  }
  return key;
};

export class Cipher {
  constructor(key) {
    if (typeof key === 'undefined') {
      this.key = generateRandomKey();
    } else if (/^[a-z]+$/.test(key)) {
      this.key = key;
    } else {
      throw new Error('Bad key');
    }
  }

  encode(str) {
    return this.xxcode(str, encoded);
  }

  decode(str) {
    return this.xxcode(str, decoded);
  }

  xxcode(str, f) {
    let coded = '';
    for (let i = 0; i < str.length; i += 1) {
      coded += f(str[i], this.keyCharAt(i));
    }
    return coded;
  }

  keyCharAt(idx) {
    return this.key.charAt(idx % this.key.length);
  }
}

File: kotlin/simple-cipher/src/main/kotlin/SimpleCipher.kt
import java.lang.Math

data class Cipher(var key: String = randomKey()) {

    companion object {
        const val ALPHABET = "abcdefghijklmnopqrstuvwxyz"
        fun randomKey(): String =
                (1..100).map { ALPHABET.random() }.joinToString("")
    }

    init {
        require(key.isNotEmpty())
        require(key.all { it in ALPHABET })
    }

    fun encode(s: String): String = code(s, +1)
    fun decode(s: String): String = code(s, -1)

    private fun code(s: String, direction: Int): String {
        require(s.all { it in ALPHABET })
        // ensure key is at least as long as the string
        while (key.length < s.length) key += key

        return s.zip(key)
                .map { codeChar(it.first, it.second, direction) }
                .joinToString("")
    }

    private fun codeChar(textChar: Char, keyChar: Char, dir: Int): Char {
        val idx = Math.floorMod(
                ALPHABET.indexOf(textChar) + dir * ALPHABET.indexOf(keyChar),
                ALPHABET.length
        )
        return ALPHABET[idx]
    }
}

